libbeat/publisher/queue/spool/config.go

```diff
@@ -0,0 +1,123 @@
+package spool
+
+import (
+	"errors"
+	"fmt"
+	"os"
+	"time"
+
+	"github.com/dustin/go-humanize"
+	"github.com/joeshaw/multierror"
+)
+
+type config struct {
+	File  pathConfig  `config:"file"`
+	Write writeConfig `config:"write"`
+	Read  readConfig  `config:"read"`
+}
+
+type pathConfig struct {
+	Path        string      `config:"path"`
+	Permissions os.FileMode `config:"permissions"`
+	MaxSize     size        `config:"size"`
+	PageSize    size        `config:"page_size"`
+	Prealloc    bool        `config:"prealloc"`
+}
+
+type writeConfig struct {
+	BufferSize   size          `config:"buffer_size"`
+	FlushTimeout time.Duration `config:"flush_timeout"`
+	Codec        codecID       `config:"codec"`
+}
+
+type readConfig struct {
+	FlushTimeout time.Duration `config:"flush_timeout"`
+}
+
+func defaultConfig() config {
+	return config{
+		File: pathConfig{
+			Path:        "",
+			Permissions: 0600,
+			MaxSize:     100 * humanize.MiByte,
+			PageSize:    4 * humanize.KiByte,
+			Prealloc:    true,
+		},
+		Write: writeConfig{
+			BufferSize:   1 * humanize.MiByte,
+			FlushTimeout: 1 * time.Second,
+			Codec:        codecCBORL,
+		},
+		Read: readConfig{
+			FlushTimeout: 0,
+		},
+	}
+}
+
+type size uint64
+
+func (c *pathConfig) Validate() error {
+	var errs multierror.Errors
+
+	if c.Path == "" {
+		errs = append(errs, errors.New("empty path"))
+	}
+
+	if c.MaxSize < humanize.MiByte {
+		errs = append(errs, errors.New("max size must be larger 1MiB"))
```

Comment:
What happens if someone has a multiline event of 2MB and the MaxSize here is 1.1MiB?

Reply:
Well, there are some "limits" that are not supported.

If event is > total file size, we will block forever.

As event sizes are dynamic and we don't want to employ hard limits on event sizes, we have kind of a weird effect here. There are potential parameters combinations + event sizes that can block the queue. The best we could do is: introduce a max event size and drop events if they hit the limit. This way we can precompute the worst page-usage of an event and figure if max size must be even larger. It's kinda tricky, that's why I didn't do it yet.